docs/design/kubeadm-api.md

```diff
@@ -0,0 +1,275 @@
+## Kubeadm Phases and API
+
+### Refactoring kubeadm
+We all want kubeadm modular and componentized, here’s the proposal for that!
+I’m not writing down all reasons for doing this; they are already available in the blog post: https://docs.google.com/document/d/1bpEGhdg8t-2Ae-IL_ezZIPklMHpZLWwnoAPMgLwT3oU
+
+### Phases
+We want to identify the phases every single Kubernetes deployment goes through in order to get the cluster up and running.
+
+In this process, I've been thinking a lot about security and tried to put that first. Also, I've read the 
+kops, kargo and kube-up code in detail to see how they are doing these things. We don't want to make an API that's incompatible with those tools.
+
+So far (for kubeadm) I've identified these:
+
+#### Certificates
+
+First, certificates should be created in a directory.
+There should be:
+ - a CA certificate (`ca.pem`) with its private key (`ca-key.pem`)
+   - Is the public key necessary to have here (`ca-pub.pem`)?
+ - an API Server certificate (`apiserver.pem`) signed with `ca.pem` as the CA with its private key (`apiserver-key.pem`)
+ - a Service Account private key that's used for signing Service Accounts' tokens.
+   - TODO: Is the certificate for the serviceaccount private key necessary?
+
+Also, we should make sure that both RSA and ECDSA private keys work with kubeadm and have tests for both.
+We probably want to expose the private key bitsize also.
```

Comment:
If we set the bitsize to be large enough, we shouldn't need to expose other options (it's just more confusing/complexity for users). If you really want a different bitsize, you should follow the "bring your own cert" path instead of letting us generate certs, as then you have complete control over the algorithm and the bitsize.